source/iopipe/json/serialize.d

```diff
@@ -971,6 +1052,128 @@ unittest
     assert(cArr.length == 3);
 }
 
+private void deserializeImplWithPolicy(T, JT, Policy)(
+    ref JT tokenizer,
+    ref T item,
+    ref Policy policy
+) if (is(T == struct) && !isInstanceOf!(JSONValue, T) && !isInstanceOf!(Nullable, T) && !__traits(hasMember, T, "fromJSON"))
+{
+    
+    
+    // Begin deserialization
+    auto visited = policy.onBegin!(JT, T)(tokenizer);
+    
+    // Process fields
+    auto jsonItem = tokenizer.nextSignificant();
+    while(true)
+    {
+        if(jsonItem.token == JSONToken.Comma)
+        {
+            jsonItem = tokenizer.nextSignificant();
+            static if(tokenizer.config.JSON5)
+            {
+                if(jsonItem.token == JSONToken.ObjectEnd)
+                    break;
+            }
+            continue;
+        }
+        
+        // Validate key
+        static if(tokenizer.config.JSON5)
+        {
+            if(jsonItem.token != JSONToken.String && jsonItem.token != JSONToken.Symbol)
+                jsonExpect(jsonItem, JSONToken.String, "Expecting member name of " ~ T.stringof);
+        }
+        else
+            jsonExpect(jsonItem, JSONToken.String, "Expecting member name of " ~ T.stringof);
+        
+        // Get field name
+        auto nameItem = jsonItem;
+        string key = nameItem.data(tokenizer.chain).to!string;
+        
+        // Expect colon
+        jsonItem = tokenizer.nextSignificant()
+            .jsonExpect(JSONToken.Colon, "Expecting colon when parsing " ~ T.stringof);
+        
+        // Let the policy handle this field
+        bool handled = policy.onField!(JT, T)(tokenizer, item, key, visited);
+        
+        if(!handled) {
+            policy.onUnknownField!(JT, T)(tokenizer, key);
+        }
+        
+        if(policy.relPol == ReleasePolicy.afterMembers)
+            tokenizer.releaseParsed();
```

Comment:
This should be handled internally (in any of the hook functions). This way, the ReleasePolicy is an internal detail of the policy (in this case, using an enum, but in some other custom policy could be done differently), and this leaves the impl function concerned about just dealing with the structure of the parsed data, and calling hooks. If we have this in here, then all policies must provide a release policy with that enumeration, and also will have to adhere to whatever we define for the enum.

What I'm getting at is that we can separate concerns here -- have a function that processes the structure of the incoming JSON data, and leave the custom handling to the policy.